Improve indentation

fpringvaldsen opened the pull request:
This PR improves the indentation when modifying values in a ConfigDocument.

fpringvaldsen commented:
Most of the significant quirks of this are called out in the tests, and it obviously won't work perfectly in the case where it gets a file with completely crazy formatting.

The only other quirk I can think of is the fact that, when indenting a setting on a value replacement, `ConfigNodeObject` uses the `indentation()` method, which finds the indentation of the first setting, so if there are other settings in the map that come before the one that matched and they have different indentations, the inserted value will be indented strangely. I'm going to work on fixing that now, as it shouldn't take long and shouldn't interfere with review of what's currently here.

fpringvaldsen commented:
Okay, this should be good to go for review. @havocp Let me know if you want me to add more test coverage!

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -24,4 +24,29 @@
+            return new ConfigNodeObject(childrenCopy);
+        } else {
+            return new ConfigNodeConcatenation(childrenCopy);
+        }
```
This if/then/else feels like it should be a virtual method

havocp commented on config/src/main/java/com/typesafe/config/impl/ConfigNodeComplexValue.java:
```diff
@@ -24,4 +24,29 @@
+    protected ConfigNodeComplexValue indentText(AbstractConfigNode indentation) {
+        ArrayList<AbstractConfigNode> childrenCopy = new ArrayList<AbstractConfigNode>(children);
+        for (int i = 0; i < childrenCopy.size(); i++) {
+            if (childrenCopy.get(i) instanceof ConfigNodeSingleToken &&
```
if you do `AbstractConfigNode child = childrenCopy.get(i)` at the top here, you'd avoid a lot of noisy `childrenCopy.get(i)` repetition

fpringvaldsen commented:
@havocp I've addressed your feedback! Also, I've noticed when it's searching for a setting from which to use the indentation, it won't look at include nodes, so I'm thinking about adding a test/functionality for that sometime later today.

fpringvaldsen commented:
Also a question that came up while working through this: I've noticed that `ConfigNodeInclude` subclasses `ConfigNodeComplexValue`. Do we want it to work that way? Or should `ConfigNodeInclude` not be an instance of a value class?

havocp commented:
I would like to factor meaningful semantic types of ConfigNode up into the public API as interfaces (ConfigNodeValue, ConfigNodeObject, ConfigNodeList, ConfigNodeInclude, ConfigNodeSubstitution, ...). Then the concrete types would be called Simple or Abstract or whatever. I was going to do this soon if you want.

fpringvaldsen commented:
@havocp No rush! I think this PR gives us everything we need for our use case. I don't think changing `ConfigNodeInclude` to not subclass `ConfigNodeComplexValue` will change the way the config document stuff works. I was mostly just wondering if we should consider an include statement a "value". If so, the current implementation makes sense, but if not, changing `ConfigNodeInclude` to not subclass `ConfigNodeComplexValue` might make the implementation clearer.

havocp commented:
if I remember right, includes expand into a bunch of fields that merge into the current object. so they are not really values.

fpringvaldsen commented:
Thanks! I think it might make sense then to not subclass `ConfigNodeComplexValue` in `ConfigNodeInclude`, so I'll make a commit to change that.

fpringvaldsen commented:
@havocp I've added that newest commit! I don't think there are any more changes I want to make on this PR, unless you have anymore feedback.

havocp commented:
Thanks!
